File: AndroidArtStudy/Chapter02/demo06/src/main/java/com/example/demo06/MessengerService.java
package com.example.demo06;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.util.Log;

public class MessengerService extends Service {

    private static final String TAG = "MessengerService";

    public static final int MSG_SERVICE = 0;
    public static final int MSG_CLIENT = 1;


    private static class MessengerHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MSG_CLIENT:
                    String str = msg.getData().getString("msg");
                    Log.d(TAG, str);
                    Messenger replyTo = msg.replyTo;
                    Message replyMsg = Message.obtain(null, MSG_SERVICE);
                    Bundle bundle = new Bundle();
                    bundle.putString("reply", "已收到: " + str);
                    replyMsg.setData(bundle);
                    try {
                        replyTo.send(replyMsg);
                    } catch (RemoteException e) {
                        e.printStackTrace();
                    }
                    break;
                default:
                    super.handleMessage(msg);
                    break;
            }
        }
    }

    private Messenger messenger = new Messenger(new MessengerHandler());


    public MessengerService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        return messenger.getBinder();
    }
}

File: AndroidArtStudy/Chapter02/demo05/src/main/java/com/example/demo05/MessengerService.kt
package com.example.demo05

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.*
import android.util.Log

/**
 * Created by liu on 2016/11/27 0027.
 */
class MessengerService : Service() {
    companion object {
        val TAG = "MessengerService"

        val MSG_FROM_CLIENT = 0
        val MSG_FROM_SERVICE = 1

    }

    private class MessengerHandler : Handler() {

        override fun handleMessage(msg: Message) {
            when (msg.what) {
                MSG_FROM_CLIENT -> {
                    Log.i(TAG, "receive msg from Client:" + msg.data.getString("msg"))
                    val client=msg.replyTo
                    val relpyNessage=Message.obtain(null, MSG_FROM_SERVICE)
                    val bundle=Bundle()
                    bundle.putString("reply","嗯，你的消息我已收到，稍后会回复你。")
                    relpyNessage.data=bundle

                    try {
                        client.send(relpyNessage)
                    } catch(e: Exception) {
                        e.printStackTrace()
                    }

                }
                else -> super.handleMessage(msg)
            }
        }
    }

    private val mMessenger = Messenger(MessengerHandler())

    override fun onBind(p0: Intent?): IBinder? {
        return mMessenger.binder
    }
}